File: leetcode/1570.py
"""
Runtime: 1672 ms, faster than 99.91% of Python3 online submissions for Dot Product of Two Sparse Vectors.
Memory Usage: 18.6 MB, less than 39.32% of Python3 online submissions for Dot Product of Two Sparse Vectors.
"""


class SparseVector:
    def __init__(self, nums: List[int]):
        self.nums = nums
        self.len = len(nums)

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: "SparseVector") -> int:
        res = 0
        for i in range(vec.len):
            if vec.nums[i] != 0 and self.nums[i] != 0:
                res += vec.nums[i] * self.nums[i]

        return res


# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector(nums1)
# v2 = SparseVector(nums2)
# ans = v1.dotProduct(v2)

File: leetcode/1570.cpp
// Runtime: 293 ms, faster than 56.61% of C++ online submissions for Dot Product of Two Sparse Vectors.
// Memory Usage: 166.2 MB, less than 85.45% of C++ online submissions for Dot Product of Two Sparse Vectors.
class SparseVector {
 private:
  // nonzero is the list of idx, value
  vector<pair<int, int>> nonzero;

 public:
  SparseVector(vector<int> &nums) {
    // record the idx, val for nonzero values
    for (vector<int>::size_type i = 0; i < nums.size(); ++i) {
      if (nums[i] != 0) {
        nonzero.push_back(pair<int, int>(i, nums[i]));
      }
    }
  }

  // Return the dotProduct of two sparse vectors
  int dotProduct(SparseVector &vec) {
    // iterator for self and vec
    int i = 0, j = 0;

    // accumulate the nonzero multiplications
    int res = 0;

    while (i < nonzero.size() && j < vec.nonzero.size()) {
      // add contribution if first values equal
      if (nonzero[i].first == vec.nonzero[j].first) {
        res += nonzero[i].second * vec.nonzero[j].second;
        ++i;
        ++j;
        continue;
      }

      // push up i and j
      while (i < nonzero.size() and j < vec.nonzero.size() and
             nonzero[i].first < vec.nonzero[j].first) {
        ++i;
      }
      while (i < nonzero.size() and j < vec.nonzero.size() and
             vec.nonzero[j].first < nonzero[i].first) {
        ++j;
      }
    }

    return res;
  }
};
